src/test/webServer.ts

```diff
@@ -0,0 +1,146 @@
+      waitForSocket(port, 100, cancellationToken),
+    ]);
+    cancellationToken.canceled = true;
+    if (portIsUsed)
+      throw new Error(`Port ${port} is used, make sure that nothing is running on the port`);
+  }
+
+  private async _startWebServer(): Promise<number> {
+    let collectPortResolve = (port: number) => { };
+    const collectPortPromise = new Promise<number>(resolve => collectPortResolve = resolve);
+    function collectPort(data: Buffer) {
+      const regExp = /http:\/\/localhost:(\d+)/.exec(data.toString());
+      if (regExp)
+        collectPortResolve(parseInt(regExp[1], 10));
+    }
+
+    this._process = spawn(this.config.command, {
+      env: {
+        ...DEFAULT_ENVIRONMENT_VARIABLES,
+        ...process.env,
+        ...this.config.env,
+      },
+      cwd: this.config.cwd,
+      shell: true,
+      // On non-windows platforms, `detached: true` makes child process a leader of a new
+      // process group, making it possible to kill child process tree with `.kill(-pid)` command.
+      // @see https://nodejs.org/api/child_process.html#child_process_options_detached
+      detached: process.platform !== 'win32',
+    });
+    this._process.stdout.pipe(newProcessLogPrefixer()).pipe(process.stdout);
+    this._process.stderr.pipe(newProcessLogPrefixer()).pipe(process.stderr);
+    if (!this.config.port)
+      this._process.stdout.on('data', collectPort);
+    let processExitedWithNonZeroStatusCodeCallback = (error: Error) => { };
+    this._processExitedWithNonZeroStatusCode = new Promise((_, reject) => processExitedWithNonZeroStatusCodeCallback = reject);
+    this._process.on('exit', code => processExitedWithNonZeroStatusCodeCallback(new Error(`WebServer was not able to start. Exit code: ${code}`)));
+    if (this.config.port)
+      return this.config.port;
+    console.log(`Starting WebServer port detection.`);
+    const detectedPort = await Promise.race([
+      this._processExitedWithNonZeroStatusCode,
+      collectPortPromise,
+    ]);
+    console.log(`Port ${detectedPort} by process '${this.config.command}' was automatically detected.`);
+    return detectedPort;
+  }
+
+  private async _waitForAvailability(port: number) {
+    const launchTimeout = this.config.timeout || 60 * 1000;
+    const cancellationToken = { canceled: false };
+    const { timedOut } = (await Promise.race([
+      raceAgainstDeadline(waitForSocket(port, 100, cancellationToken), launchTimeout + monotonicTime()),
+      this._processExitedWithNonZeroStatusCode,
+    ]));
+    cancellationToken.canceled = true;
+    if (timedOut)
+      throw new Error(`failed to start web server on port ${port} via "${this.config.command}"`);
+  }
+  public async kill() {
+    assert(this._process);
+    if (this._process.exitCode !== null || this._process.killed)
+      return;
+    const waitForExit = new Promise(resolve => this._process?.on('exit',resolve));
+    killProcessGroup(this._process.pid);
```

Comment:
I think you'll need all the facilities of processLauncher, e.g. sync kill on process.exit and other such stuff. Let's follow up.

Reply:
sounds good, I adjusted it so it uses launchProcess now.